lsp/nls/src/requests/completion.rs

```diff
@@ -1,16 +1,227 @@
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            })
+        }
+        _ if item.id == id => match item.kind {
+            TermKind::Declaration(_, _, ValueState::Known(new_id))
+            | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                find_record_contract_fields(&linearization, new_id)
+            }
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+/// Extract identifiers from a row type which forms a record.
+/// Ensure that ty is really a row type.
+fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+    ty.iter_as_rows()
+        .filter_map(|item| match item {
+            RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+            RowIteratorItem::Tail(..) => None,
+        })
+        .collect::<Vec<_>>()
+}
+
+lazy_static! {
+    // unwrap is safe here because we know this is a correct regex.
+    // This regexp must be the reverse of the regexp in the lexer (nickel_lang::parser::lexer)
+    // to correctly parse identifiers.
+    static ref RE: regex::Regex = regex::Regex::new(r"[_a-zA-Z0-9-']*[a-zA-Z]_?").unwrap();
+}
+
+/// Get the identifier before the `.`, for record completion.
+fn get_identifier_before_dot(text: &str) -> Option<String> {
+    let text: String = text
+        .chars()
+        .rev()
+        .skip_while(|c| *c == '.') // skip . (if any)
+        .collect();
+
+    let name = RE.find(&text)?;
+    Some(name.as_str().chars().rev().collect())
+}
+
+/// Get the identifier before `.<text>` for record completion.
+fn get_identifier_before_field(text: &str) -> Option<String> {
+    // Skip `<text>`
+    let text: String = text
+        .chars()
+        .rev()
+        .skip_while(|c| RE.is_match(c.to_string().as_ref()))
+        .collect::<String>()
+        .chars()
+        .rev()
+        .collect();
+
+    get_identifier_before_dot(&text[..])
+}
+
+fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
+    let mut seen: Vec<CompletionItem> = Vec::new();
+    for item in items {
+        if !seen.iter().any(|seen_item| seen_item.label == item.label) {
+            seen.push(item.clone())
+        }
+    }
+    seen
+}
+
+/// Search the linearization to find the record information associated with a
+/// partiular ID, and in the scope of a given linearization item.
+fn collect_record_info(
+    linearization: &Completed,
+    item: &LinearizationItem<Types>,
+    name: usize,
+) -> Vec<(Vec<Ident>, Types)> {
+    linearization
+        .get_in_scope(item)
+        .iter()
+        .filter_map(|item| {
+            let (ty, _) = linearization.resolve_item_type_meta(item);
+            match (&item.kind, ty) {
+                // Get record fields from static type info
+                (TermKind::Declaration(..), Types(AbsType::Record(row))) if name == item.id => {
+                    Some((extract_ident(&row), item.ty.clone()))
```

Comment:
Why do you need to `clone` here if `extract_ident` is taking a reference?

Reply:
This `clone` is for the type, not the boxed row type that is passed as an argument to `extract_ident`.

Reply:
Ah yes, sorry, I misread the parentheses.